django/test/utils.py

```diff
@@ -401,18 +403,31 @@ def enable(self):
         self.wrapped = settings._wrapped
         settings._wrapped = override
         for key, new_value in self.options.items():
-            setting_changed.send(sender=settings._wrapped.__class__,
-                                 setting=key, value=new_value, enter=True)
+            try:
+                setting_changed.send(sender=settings._wrapped.__class__,
+                                     setting=key, value=new_value, enter=True)
+            except Exception as ex:
+                self.enable_exception = ex
+                self.disable()
 
     def disable(self):
         if 'INSTALLED_APPS' in self.options:
             apps.unset_installed_apps()
         settings._wrapped = self.wrapped
         del self.wrapped
+        responses = []
         for key in self.options:
             new_value = getattr(settings, key, None)
-            setting_changed.send(sender=settings._wrapped.__class__,
-                                 setting=key, value=new_value, enter=False)
+            responses_for_setting = setting_changed.send_robust(sender=settings._wrapped.__class__,
+                                                                setting=key, value=new_value, enter=False)
+            responses.extend(responses_for_setting)
+
+        if self.enable_exception is not None:
+            raise self.enable_exception
```

Comment:
This should be reset to `None` before raising else reusing the instance will systematically raise here. Adding a test for that (`override_settings` instance raising an exception on its first `__enter__`) wouldn't hurt as that's a common use case when decorating a test class.

Reply:
Good catch, thanks. I wasn't sure how should I recreate such situation and went with something like this: 51a25e0482b1833